src/server/req_delete.c

```diff
@@ -1120,6 +1120,12 @@ req_deleteReservation(struct batch_request *preq)
 
 			pnxj = (job *) GET_NEXT(pjob->ji_jobque);
 
+			/* skip all expired subjobs, expired subjobs are deleted when array parent is
+			 * issued delete request
+			 */
+			for (; pnxj != NULL && (pnxj->ji_qs.ji_svrflags & JOB_SVFLG_SubJob) &&
```

Comment:
Aren't ALL subjobs deleted when an array parent is deleted? So shouldn't you skip over all subjobs regardless of their state?

Reply:
One more thing, can there ever be a case where subjobs of an arrayjob are not listed all together? Suppose the arrayjob's user exceeds some limit after 2 of the subjobs were run, so the scheduler runs 2 of the subjobs and some other job. Then later more subjobs are run. Since subjobs aren't real jobs until they get run, can the subjobs possibly be listed non-consecutively in ji_jobque in such a case?

Reply:
Regarding your first query, all subjobs (expired and running are deleted) when a parent job is deleted. expired subjobs are deleted instantaneously but for running subjobs it goes through a different code path of sending a signal to mom to delete the running job and then delete the array parent when all the subjobs are deleted.
For your second comment, As long as array parent is part of the reservation queue's job list, it should be fine. deletion of parent job should trigger delete of all expired subjobs which will make them dequeue from reservation's job queue and the server will not see those pointers anymore in the req_deleteReservation function.

Reply:
Since we already skip sending any subjob for deletion  (line 1135) I do not need the block to check pjob is null. I'll remove that block.

Reply:
"Regarding your first query, all subjobs (expired and running are deleted) when a parent job is deleted. expired subjobs are deleted instantaneously but for running subjobs it goes through a different code path of sending a signal to mom to delete the running job and then delete the array parent when all the subjobs are deleted."

So, if we only skip over the expired jobs, then we might set next pointer to a queued/running subjob. In such a case, isn't it possible that by the time we access the next pointer, the queued/running subjob has been deleted and the pointer is invalid? Or what about the case where the list looks like this: parent job->expired subjob->a qrun job->a running subjob, now, isn't it possible that we set the next pointer to the running subjob and by the time we dereference it, the running subjob gets deleted and the pointer is invalid?

Reply:
We are actually safe in case of queued subjobs because subjobs are enqueued when they start running. Now if the next pointer is pointing to the running subjob then that should be ok too because running subjobs are not deleted instantaneously. 
Now for the case related to parent job->expired subjob->qrun job-> running subjob. In this case, what will happen is that server will try to delete parent job and as part of that it will delete all the expired subjobs and do an issue_signal to moms for running subjobs and register a Deferred_Reply task in case mom replies back. Since the server is single threaded I think we are safe here too.